File: Lab_8/Main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

// Clase Abstracta Vehiculo
abstract class Vehiculo {
    protected String marca;
    protected String modelo;
    protected int año;

    public Vehiculo(String marca, String modelo, int año) {
        this.marca = marca;
        this.modelo = modelo;
        this.año = año;
    }

    public abstract void encender();
    public abstract void apagar();
    public abstract void describir();
}

// Subclase Auto
class Auto extends Vehiculo {
    public Auto(String marca, String modelo, int año) {
        super(marca, modelo, año);
    }

    @Override
    public void encender() {
        System.out.println("El auto esta encendiendo.");
    }

    @Override
    public void apagar() {
        System.out.println("El auto se ha apagado.\n");
    }

    @Override
    public void describir() {
        System.out.println("Este es un auto de marca " + marca + ", modelo " + modelo + ", del año " + año + ".");
    }
}

// Subclase Motocicleta
class Motocicleta extends Vehiculo {
    public Motocicleta(String marca, String modelo, int año) {
        super(marca, modelo, año);
    }

    @Override
    public void encender() {
        System.out.println("La motocicleta esta encendiendo.");
    }

    @Override
    public void apagar() {
        System.out.println("La motocicleta se ha apagado.\n");
    }

    @Override
    public void describir() {
        System.out.println("Esta es una motocicleta de marca " + marca + ", modelo " + modelo + ", del año " + año + ".");
    }
}

// Subclase Camion
class Camion extends Vehiculo {
    public Camion(String marca, String modelo, int año) {
        super(marca, modelo, año);
    }

    @Override
    public void encender() {
        System.out.println("El camion esta encendiendo.");
    }

    @Override
    public void apagar() {
        System.out.println("El camion se ha apagado.\n");
    }

    @Override
    public void describir() {
        System.out.println("Este es un camion de marca " + marca + ", modelo " + modelo + ", del año " + año + ".");
    }
}

// Programa Principal
public class Main {
    public static void main(String[] args) {
        Scanner lector = new Scanner(System.in);
        List<Vehiculo> vehiculos = new ArrayList<>();

        while (true) {
            System.out.println("\nSeleccione el tipo de vehiculo a agregar:");
            System.out.println("1. Auto");
            System.out.println("2. Motocicleta");
            System.out.println("3. Camion");
            System.out.println("4. Salir");
            int opcion = lector.nextInt();
            lector.nextLine(); // Limpiar el buffer

            if (opcion == 4) {
                break;
            }

            System.out.print("Ingrese la marca: ");
            String marca = lector.nextLine();
            System.out.print("Ingrese el modelo: ");
            String modelo = lector.nextLine();
            System.out.print("Ingrese el año: ");
            int año = lector.nextInt();
            lector.nextLine(); // Limpiar el buffer

            switch (opcion) {
                case 1:
                    vehiculos.add(new Auto(marca, modelo, año));
                    break;
                case 2:
                    vehiculos.add(new Motocicleta(marca, modelo, año));
                    break;
                case 3:
                    vehiculos.add(new Camion(marca, modelo, año));
                    break;
                default:
                    System.out.println("Opcion no valida.");
            }
        }

        System.out.println("\n--- Mostrando informacion de los vehiculos ---");
        for (Vehiculo vehiculo : vehiculos) {
            vehiculo.encender();
            vehiculo.describir();
            vehiculo.apagar();
        }

        lector.close();
    }
}

File: Lab_8/Main.py
from abc import ABC, abstractmethod

# Clase Abstracta Vehiculo
class Vehiculo(ABC):
    def __init__(self, marca, modelo, año):
        self.marca = marca
        self.modelo = modelo
        self.año = año

    @abstractmethod
    def encender(self):
        pass

    @abstractmethod
    def apagar(self):
        pass

    @abstractmethod
    def describir(self):
        pass


# Subclase Auto
class Auto(Vehiculo):
    def encender(self):
        print("El auto esta encendiendo.")

    def apagar(self):
        print("El auto se ha apagado.\n")

    def describir(self):
        print(f"Este es un auto de marca {self.marca}, modelo {self.modelo}, del año {self.año}.")


# Subclase Motocicleta
class Motocicleta(Vehiculo):
    def encender(self):
        print("La motocicleta esta encendiendo.")

    def apagar(self):
        print("La motocicleta se ha apagado.\n")

    def describir(self):
        print(f"Esta es una motocicleta de marca {self.marca}, modelo {self.modelo}, del año {self.año}.")


# Subclase Camion
class Camion(Vehiculo):
    def encender(self):
        print("El camion esta encendiendo.")

    def apagar(self):
        print("El camión se ha apagado.\n")

    def describir(self):
        print(f"Este es un camion de marca {self.marca}, modelo {self.modelo}, del año {self.año}.")


# Programa Principal
def main():
    vehiculos = []

    while True:
        print("\nSeleccione el tipo de vehiculo a agregar:")
        print("1. Auto")
        print("2. Motocicleta")
        print("3. Camion")
        print("4. Salir")

        opcion = int(input("Opcion: "))

        if opcion == 4:
            break

        if opcion not in [1, 2, 3]:
            print("Opcion no valida. Intente nuevamente.")
            continue

        marca = input("Ingrese la marca: ")
        modelo = input("Ingrese el modelo: ")
        año = int(input("Ingrese el año: "))

        if opcion == 1:
            vehiculos.append(Auto(marca, modelo, año))
        elif opcion == 2:
            vehiculos.append(Motocicleta(marca, modelo, año))
        elif opcion == 3:
            vehiculos.append(Camion(marca, modelo, año))

    print("\n--- Mostrando informacion de los vehiculos ---")
    for vehiculo in vehiculos:
        vehiculo.encender()
        vehiculo.describir()
        vehiculo.apagar()

if __name__ == "__main__":
    main()
